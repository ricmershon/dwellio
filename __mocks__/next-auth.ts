export const getServerSession = jest.fn()
export const signIn = jest.fn()
export const signOut = jest.fn()

export const useSession = jest.fn(() => ({
    data: null,
    status: 'unauthenticated'
}))

export const SessionProvider = ({ children }: { children: React.ReactNode }) => children

const nextAuth = {
    getServerSession,
    signIn,
    signOut,
    useSession,
    SessionProvider
}

export default nextAuth