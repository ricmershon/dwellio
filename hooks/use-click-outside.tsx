import { RefObject, useEffect } from "react";

type AnyEvent = MouseEvent | TouchEvent | PointerEvent | KeyboardEvent;

export const useClickOutside = (
    refs: RefObject<HTMLElement | null> | RefObject<HTMLElement | null>[], // single or multiple refs considered "inside"
    onOutside: (event: AnyEvent) => void,
    enabled = true
) => {
    useEffect(() => {
        if (!refs) {
            return;
        }

        const refList = Array.isArray(refs) ? refs : [refs];
        if (!enabled) {
            return;
        }

        const isInside = (target: EventTarget | null) => {
            const node = target as Node | null;
            return refList.some((ref) => ref.current && (ref.current === node || ref.current.contains(node)));
        };

        const onPointer = (event: PointerEvent) => {
            if (!isInside(event.target)) {
                onOutside(event);
            }
        };

        const onKey = (event: KeyboardEvent) => {
            if (event.key === "Escape") {
                onOutside(event);
            }
        };

        // pointerdown fires earlier than click; feels snappier and covers mouse+touch+pen
        document.addEventListener("pointerdown", onPointer, true); // capture to beat stopPropagation
        document.addEventListener("keydown", onKey);

        return () => {
            document.removeEventListener("pointerdown", onPointer, true);
            document.removeEventListener("keydown", onKey);
        };
    }, [enabled, onOutside, refs]);
}
