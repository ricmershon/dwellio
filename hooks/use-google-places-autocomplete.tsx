import { useEffect, useState } from "react";

import { fetchPlacesAutocomplete } from "@/lib/places-autocomplete";
import { AutocompleteFetchOptions, AutocompletePrediction } from "@/types/types";

export function usePlacesAutocomplete(query: string) {
    const [predictions, setPredictions] = useState<AutocompletePrediction[]>([]);

    useEffect(() => {
        if (!query) {
            setPredictions([]);
            return;
        }

        const abortController = new AbortController();

        fetchPlacesAutocomplete(query, {
            signal: abortController.signal,
            includeDetails: true,
            maxDetails: 5,
            regionCode: "US",
            includedPrimaryTypes: ["street_address"],
        } as AutocompleteFetchOptions)
            .then((response) => {
                if (!abortController.signal.aborted) {
                    setPredictions(response)
                }
            })
            .catch((error) => {
                if (!abortController.signal.aborted) {
                    console.error("Google Maps Places Autocomplete error:", error);
                    setPredictions([]);
                }
            });

        return () => abortController.abort();
    }, [query]);

    return { predictions };
}
