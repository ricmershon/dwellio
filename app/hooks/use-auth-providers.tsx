import { useEffect, useState } from 'react';
import { getProviders, ClientSafeProvider, LiteralUnion } from 'next-auth/react';
import { BuiltInProviderType } from "next-auth/providers/index";

export const useAuthProviders = () => {
    const [providers, setProviders] = useState<
        Record<LiteralUnion<BuiltInProviderType, string>, ClientSafeProvider> | null
    >(null);

    useEffect(() => {
        const setAuthProviders = async () => {
            const response = await getProviders();
            setProviders(response);
        };
        setAuthProviders();
    }, []);

    return providers;
};