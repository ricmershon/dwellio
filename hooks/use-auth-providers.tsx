import { useEffect, useState } from "react";
import { getProviders, ClientSafeProvider, LiteralUnion } from "next-auth/react";
import { BuiltInProviderType } from "next-auth/providers/index";

export const useAuthProviders = () => {
    const [providers, setProviders] = useState<
        Record<LiteralUnion<BuiltInProviderType, string>, ClientSafeProvider> | null
    >(null);

    useEffect(() => {
        getProviders().then((response) => setProviders(response));
    }, []);

    return providers;
};