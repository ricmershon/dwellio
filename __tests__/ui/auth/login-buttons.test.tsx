// Mock Next.js navigation
jest.mock('next/navigation', () => {
	const mockPush = jest.fn();
	const mockReplace = jest.fn();
	const mockBack = jest.fn();
	const mockForward = jest.fn();
	const mockRefresh = jest.fn();
	const mockPrefetch = jest.fn();
	const mockSearchParams = new URLSearchParams();

	return {
		usePathname: jest.fn(() => '/'),
		useSearchParams: jest.fn(() => mockSearchParams),
		useRouter: jest.fn(() => ({
			push: mockPush,
			replace: mockReplace,
			back: mockBack,
			forward: mockForward,
			refresh: mockRefresh,
			prefetch: mockPrefetch,
		})),
		useParams: jest.fn(() => ({})),
		// Export mocks for testing
		mockPush,
		mockReplace,
		mockBack,
		mockForward,
		mockRefresh,
		mockPrefetch,
		mockSearchParams,
	};
});

// Mock NextAuth
jest.mock('next-auth/react', () => ({
	signIn: jest.fn(),
}));

import React from 'react';
import { screen, fireEvent } from '@testing-library/react';
import { signIn, ClientSafeProvider, LiteralUnion } from 'next-auth/react';
import { BuiltInProviderType } from 'next-auth/providers/index';
import { useSearchParams } from 'next/navigation';

import LoginButtons from '@/ui/login/login-buttons';
import { useAuthProviders } from '@/hooks/use-auth-providers';
import { render, createMockSearchParams } from '@/__tests__/test-utils';

// Mock custom hook
jest.mock('@/hooks/use-auth-providers', () => ({
	useAuthProviders: jest.fn(),
}));

describe('LoginButtons', () => {
	const mockSignIn = signIn as jest.MockedFunction<typeof signIn>;
	const mockUseSearchParams = useSearchParams as jest.MockedFunction<typeof useSearchParams>;
	const mockUseAuthProviders = useAuthProviders as jest.MockedFunction<typeof useAuthProviders>;

	const mockSearchParams = createMockSearchParams();

	const mockProviders: Partial<Record<LiteralUnion<BuiltInProviderType, string>, ClientSafeProvider>> = {
		google: {
			id: 'google',
			name: 'Google',
			type: 'oauth',
			signinUrl: 'http://localhost:3000/api/auth/signin/google',
			callbackUrl: 'http://localhost:3000/api/auth/callback/google',
		} as ClientSafeProvider,
		github: {
			id: 'github',
			name: 'GitHub',
			type: 'oauth',
			signinUrl: 'http://localhost:3000/api/auth/signin/github',
			callbackUrl: 'http://localhost:3000/api/auth/callback/github',
		} as ClientSafeProvider,
	};

	beforeEach(() => {
		jest.clearAllMocks();
		mockUseSearchParams.mockReturnValue(mockSearchParams);
		(mockSearchParams.get as jest.Mock).mockReturnValue(null); // Default: no returnTo param
		// eslint-disable-next-line @typescript-eslint/no-explicit-any
		mockUseAuthProviders.mockReturnValue(mockProviders as any);
	});

	describe('Rendering', () => {
		it('should render login buttons for all available providers', () => {
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			expect(buttons).toHaveLength(2);
			expect(screen.getAllByText('Login')).toHaveLength(2);
		});

		it('should render nothing when providers are not loaded', () => {
			mockUseAuthProviders.mockReturnValue(null);
			
			const { container } = render(<LoginButtons />);
			expect(container.firstChild).toBeNull();
		});

		it('should render custom text when provided', () => {
			render(<LoginButtons text="Sign In" />);

			expect(screen.getAllByText('Sign In')).toHaveLength(2);
		});

		it('should apply custom className when provided', () => {
			render(<LoginButtons buttonClassName="custom-btn" />);

			const buttons = screen.getAllByRole('button');
			buttons.forEach(button => {
				expect(button).toHaveClass('custom-btn');
			});
		});

		it('should render custom icon when provided', () => {
			const customIcon = <span data-testid="custom-icon">üîê</span>;
			render(<LoginButtons icon={customIcon} />);

			expect(screen.getAllByTestId('custom-icon')).toHaveLength(2);
		});
	});

	describe('User Interactions', () => {
		it('should call signIn with Google provider when Google button is clicked', () => {
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			fireEvent.click(buttons[0]); // First button (Google)

			expect(mockSignIn).toHaveBeenCalledWith('google', { callbackUrl: '/' });
		});

		it('should call signIn with GitHub provider when GitHub button is clicked', () => {
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			fireEvent.click(buttons[1]); // Second button (GitHub)

			expect(mockSignIn).toHaveBeenCalledWith('github', { callbackUrl: '/' });
		});

		it('should use returnTo parameter from URL when present', () => {
			(mockSearchParams.get as jest.Mock).mockReturnValue('/profile');
			
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			fireEvent.click(buttons[0]);

			expect(mockSignIn).toHaveBeenCalledWith('google', { callbackUrl: '/profile' });
		});

		it('should default to "/" when no returnTo parameter', () => {
			(mockSearchParams.get as jest.Mock).mockReturnValue(null);
			
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			fireEvent.click(buttons[0]);

			expect(mockSignIn).toHaveBeenCalledWith('google', { callbackUrl: '/' });
		});
	});

	describe('URL Parameters', () => {
		it('should check for returnTo search parameter', () => {
			render(<LoginButtons />);

			expect(mockSearchParams.get).toHaveBeenCalledWith('returnTo');
		});

		it('should handle empty returnTo parameter', () => {
			(mockSearchParams.get as jest.Mock).mockReturnValue('');
			
			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			fireEvent.click(buttons[0]);

			expect(mockSignIn).toHaveBeenCalledWith('google', { callbackUrl: '/' });
		});
	});

	describe('Provider Variations', () => {
		it('should handle single provider', () => {
			const singleProvider = {
				google: mockProviders.google,
			};
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			mockUseAuthProviders.mockReturnValue(singleProvider as any);

			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			expect(buttons).toHaveLength(1);
		});

		it('should handle multiple providers', () => {
			const extendedProviders = {
				...mockProviders,
				discord: {
					id: 'discord',
					name: 'Discord',
					type: 'oauth',
					signinUrl: 'http://localhost:3000/api/auth/signin/discord',
					callbackUrl: 'http://localhost:3000/api/auth/callback/discord',
				} as ClientSafeProvider,
			};
			
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			mockUseAuthProviders.mockReturnValue(extendedProviders as any);

			render(<LoginButtons />);

			const buttons = screen.getAllByRole('button');
			expect(buttons).toHaveLength(3);
		});
	});

	describe('Snapshots', () => {
		it('should match snapshot in different states', () => {
			// Default state
			const { container } = render(<LoginButtons />);
			expect(container.firstChild).toMatchSnapshot('default state');
			
			// Null providers state
			mockUseAuthProviders.mockReturnValue(null);
			const { container: nullContainer } = render(<LoginButtons />);
			expect(nullContainer.firstChild).toMatchSnapshot('null providers');
		});
	});

});