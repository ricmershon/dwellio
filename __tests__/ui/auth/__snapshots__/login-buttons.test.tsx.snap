// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoginButtons Snapshots should match snapshot in different states: default state 1`] = `
<button
  class=""
>
  <span>
    Login
  </span>
</button>
`;

exports[`LoginButtons Snapshots should match snapshot in different states: null providers 1`] = `null`;
